lib/Sema/CSSimplify.cpp

```diff
@@ -5578,25 +5578,47 @@ bool ConstraintSystem::repairFailures(
     return false;
   }
 
+  // Be a little lax in protocol requirement matching.
   if (auto *VD = getAsDecl<ValueDecl>(anchor)) {
-    // Matching a witness to a ObjC protocol requirement.
-    if (VD->isObjC() && VD->isProtocolRequirement() &&
-        path[0].is<LocatorPathElt::Witness>() && 
+    if (VD->isProtocolRequirement() &&
+        path[0].is<LocatorPathElt::Witness>() &&
         // Note that the condition below is very important,
         // we need to wait until the very last moment to strip
         // the concurrency annotations from the inner most type.
         conversionsOrFixes.empty()) {
-      // Allow requirements to introduce `swift_attr` annotations
-      // (note that `swift_attr` in type contexts weren't supported
-      // before) and for witnesses to adopt them gradually by matching
-      // with a warning in non-strict concurrency mode.
-      if (!(Context.isSwiftVersionAtLeast(6) ||
+      bool hasAdjustments = false;
+      Type strippedLHS = lhs;
+      Type strippedRHS = rhs;
+
+      // When matching to an ObjC protocol requirement, allow requirements
+      // to introduce `swift_attr` annotations (note that `swift_attr` in
+      // type contexts weren't supported before) and for witnesses to
+      // adopt them gradually by matching with a warning in non-strict
+      // concurrency mode.
+      if (VD->isObjC() &&
+          !(Context.isSwiftVersionAtLeast(6) ||
             Context.LangOpts.StrictConcurrencyLevel ==
                 StrictConcurrency::Complete)) {
-        auto strippedLHS = lhs->stripConcurrency(/*resursive=*/true,
-                                                 /*dropGlobalActor=*/true);
-        auto strippedRHS = rhs->stripConcurrency(/*resursive=*/true,
-                                                 /*dropGlobalActor=*/true);
+        strippedLHS = lhs->stripConcurrency(/*recursive=*/true,
+                                            /*dropGlobalActor=*/true);
+        strippedRHS = rhs->stripConcurrency(/*recursive=*/true,
+                                            /*dropGlobalActor=*/true);
+        hasAdjustments = true;
+      }
+
+      // Always allow @isolated(any) mismatches.
+      auto maybeRemoveIsolatedAny = [&](Type &ty) {
+        if (auto fnType = ty->getAs<AnyFunctionType>();
+              fnType && fnType->getIsolation().isErased()) {
+          ty = fnType->withExtInfo(fnType->getExtInfo()
+                 .withIsolation(FunctionTypeIsolation::forNonIsolated()));
+          hasAdjustments = true;
+        }
+      };
+      maybeRemoveIsolatedAny(strippedLHS);
+      maybeRemoveIsolatedAny(strippedRHS);
```

Comment:
This means that `isolated(any)` requirement could be witnessed by a function type without `isolated(any)` is that, okay? Seems like it might be worth to add a test for that...

Reply:
Requirements can't themselves be `@isolated(any)`, only function types.  I do have a test for a requirement that *takes* an `@isolated(any)` function, though.

Reply:
What I meant is - is it correct for a requirement have a function type parameter that is `isolated(any)` which is then witnessed by a corresponding function type in the witness parameter that doesn’t have that isolation? You are stripping this from both sides here and there is no new test that checks this case.

Reply:
As in:

```
protocol Enqueuer {
  associatedtype Result
  func enqueue(operation: @escaping @isolated(any) () async -> Result)
}

struct A<T> : Enqueuer {
  func enqueue(operation: @escaping () async -> T) {}
}
```

Reply:
I did have a test case that tested that, actually.  But I've decided not to be permissive here; it seems unnecessarily confusing (and would need extra restrictions, e.g. around `async`) and shouldn't be necessary for source compatibility anyway.